compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -0,0 +1,312 @@
+                instrumentLifted(transformed)
+              else
+                val transformed = cpy.Apply(tree)(transformedFun, transform(args))
+                instrument(transformed)
+            else
+              tree
+
+          // f(args)
+          case tree: Apply =>
+            if canInstrumentApply(tree) then
+              if needsLift(tree) then
+                instrumentLifted(tree)
+              else
+                instrument(super.transform(tree))
+            else
+              tree
+
+          // (f(x))[args]
+          case TypeApply(fun: Apply, args) =>
+            cpy.TypeApply(tree)(coverageTransform(fun), args)
+
+          // a.b
+          case Select(qual, name) =>
+            if qual.symbol.exists && qual.symbol.is(JavaDefined) then
+              // Java classes can't be used as a value, we can't instrument the
+              // qualifier ({<Probe>;System}.xyz() is not possible !), so instrument it
+              // as it is
+              instrument(tree)
+            else
+              val transformed = cpy.Select(tree)(coverageTransform(qual), name)
+              if transformed.qualifier.isDef then
+                // instrument calls to methods without parameter list
+                instrument(transformed)
+              else
+                transformed
+
+          case tree @ Match(sel, cases) =>
+            // force the cases if needed
+            cpy.Match(tree)(coverageTransform(sel), transformCases(cases, forceInstrument))
+
+          case tree: CaseDef =>
+            transformCaseDef(tree, forceInstrument)
+
+          case tree: ValDef =>
+            // only transform the rhs
+            val rhs = coverageTransform(tree.rhs)
+            cpy.ValDef(tree)(rhs=rhs)
+
+          case tree: DefDef =>
+            // only transform the params (for the default values) and the rhs
+            // force instrumentation of literals and other small trees in the rhs,
+            // to ensure that the method call are recorded
+            val paramss = transformParamss(tree.paramss)
+            val rhs = coverageTransform(tree.rhs, forceInstrument=true)
+            cpy.DefDef(tree)(tree.name, paramss, tree.tpt, rhs)
+
+          case tree: PackageDef =>
+            // only transform the statements of the package
+            cpy.PackageDef(tree)(tree.pid, transform(tree.stats))
+          case tree: Assign =>
+            // only transform the rhs
+            cpy.Assign(tree)(tree.lhs, coverageTransform(tree.rhs))
+
+          case tree: Block =>
+            val blockCtx =
+              if forceInstrument then
+                // propagates forceInstrument with a context property, because Block transformation
+                // is handled by the superclass in transformBlock (which will call our transformBlockExpr).
+                ctx.fresh.setProperty(ForceInstrumentBlockExpr, true)
```

Comment:
It seems like it'd be simpler to always use this context property and remove the forceInstrument argument.

Reply:
Because it's false by default and not propagated in many cases, I would have to unset the property everywhere it's not used (or for all, but it would be a bit wasteful to unset it just to set it later)...

Reply:
Ah, I missed that  it wasn't propagate everywhere. I'm not sure I get the logic here, why do we propagate it to the cases of a Try expression for example?

Reply:
expr and finalizer are always guaranteed to be instrumented thanks to the `instrument` call. But cases isn't, and if it consists of a literal it won't be instrumented.

Reply:
While explaining it I'm thinking that it would be better to always force the catch cases of try blocks, actually ^^

Reply:
Added 1 commit :)

Note that it shouldn't propagate in general, because in
```scala
def f = g(0)
```
the ideal is to instrument `g` but not `0`. Therefore, the case `Apply` doesn't propagate the flag.